include/deal.II/lac/trilinos_tpetra_sparse_matrix.templates.h

```diff
@@ -1005,6 +1005,68 @@ namespace LinearAlgebra
 
 
 
+    template <typename Number, typename MemorySpace>
+    void
+    SparseMatrix<Number, MemorySpace>::clear_row(const size_type row,
+                                                 const Number    new_diag_value)
+    {
+      clear_rows(std::vector<size_type>(1, row), new_diag_value);
+    }
+
+
+
+    template <typename Number, typename MemorySpace>
+    void
+    SparseMatrix<Number, MemorySpace>::clear_rows(
+      const std::vector<size_type> &rows,
+      const Number                  new_diag_value)
+    {
+      // If the matrix is marked as compressed, we need to
+      // call resumeFill() first.
+      if (compressed || matrix->isFillComplete())
+        {
+          matrix->resumeFill();
+          compressed = false;
+        }
+
+      std::vector<int>    col_indices_vector;
+      std::vector<Number> values_vector;
+
+      for (size_type row : rows)
+        {
+          // Only do this on the rows owned locally on this processor.
+          int local_row = matrix->getRowMap()->getLocalElement(row);
+          if (local_row != Teuchos::OrdinalTraits<int>::invalid())
+            {
+              size_t nnz = matrix->getNumEntriesInLocalRow(local_row);
```

Comment:
```suggestion
              const size_t nnz = matrix->getNumEntriesInLocalRow(local_row);
```

Reply:
Can't do that. We extract the real number of non-zero entries in `getLocalRowCopy` but use the value set here to make sure we have allocated enough memory.